scylla/src/transport/topology.rs

```diff
@@ -870,6 +871,106 @@ async fn query_user_defined_types(
+
+    // Build an indegree map: for each node in the graph, how many directly depending types it has.
+    let mut indegree = defs
+        .into_iter()
+        .map(|def| {
+            (
+                (def.keyspace_name.clone(), def.type_name.clone()),
+                (def, Cell::new(0u32)),
+            )
+        })
+        .collect::<HashMap<_, _>>();
+
+    // For each node in the graph...
+    for (def, _) in indegree.values() {
+        let mut increment_referred_udts = |type_name: &str| {
+            let deg_cell = indegree
+                .get(&(def.keyspace_name.clone(), type_name.to_string()))
+                .map(|(_, count)| count)
+                .unwrap();
+            deg_cell.set(deg_cell.get() + 1);
+        };
+        // For each type referred by the node...
+        for field_type in def.field_types.iter() {
+            let pre_cql_type = map_string_to_cql_type(field_type);
+            // If the type referred is another node in the graph...
+            if let Ok(type_) = pre_cql_type {
+                do_with_referenced_udts(&mut increment_referred_udts, &type_);
+            }
+        }
+    }
+
+    let mut leaves_first = vec![];
+    // The "dual" hash map serves as a temporary storage in order to overcome the limitation of drain_filter() being unstable.
+    let mut indegree_dual = HashMap::with_capacity(indegree.len());
+
+    // While there are any nodes remaining in the graph...
+    while !indegree.is_empty() {
+        let extend_beg = leaves_first.len();
+
+        // Remove nodes with indegree zero (i.e. leaves)
+        for (key, (def, deg)) in indegree.drain() {
+            if deg.get() == 0 {
+                leaves_first.push(def);
+            } else {
+                indegree_dual.insert(key, (def, deg));
+            }
+        }
+
+        if leaves_first.len() == extend_beg {
+            panic!(
+                "Bug in fetched User Defined Types definitions: circular type dependency detected.
+                 Topological sort is thus impossible. Panicking to prevent an infinite loop."
+            )
+        }
+
+        mem::swap(&mut indegree, &mut indegree_dual);
+
+        for new_leaf in &leaves_first[extend_beg..] {
+            let mut decrement_referred_udts = |type_name: &str| {
+                *indegree
+                    .get_mut(&(new_leaf.keyspace_name.clone(), type_name.to_string()))
+                    .unwrap()
+                    .1
+                    .get_mut() -= 1;
+            };
+            // For each node that the currently removed node was dependant upon...
+            for field_type in new_leaf.field_types.iter() {
+                let pre_cql_type = map_string_to_cql_type(field_type);
+                if let Ok(type_) = pre_cql_type {
+                    do_with_referenced_udts(&mut decrement_referred_udts, &type_);
+                }
+            }
+        }
+    }
+
+    leaves_first.reverse();
+    mem::swap(&mut leaves_first, udts);
+}
```

Comment:
I'm not happy that the toposort algorithm is quadratic.

I think you are trying too hard to remove values from the map soon after their counter is decremented to zero. I don't think you have to do that - you can just operate on the hashmap keys.

Let's try this: keep doing what you did before up until line `925`, then something like this should follow (didn't compile it):

```rust
let mut sorted = Vec::with_capacity(indegree.len());
let mut next_idx = 0;

// Schedule keys that had an initial indeg of 0
for (key, (_, deg)) in indegree.iter() {
    if deg.get() == 0 {
        scheduled.push(&key);
    }
}

while let Some(key) = sorted.get(next_idx) {
    next_idx += 1;
    // Decrement the counters of all UDTs that this UDT depends upon
    // and then schedule them if their counter drops to 0
    let mut decrement_referred_udts = |type_name: &str| {
        let cnt = &indegree
            .get(key)
            .unwrap()
            .1;
        let new_cnt = cnt.get() - 1;
        cnt.set(new_cnt);
        if new_cnt == 0 {
            scheduled.push(key);
        }
    };
    let def = indegs.get(key).unwrap().0;
    for field_type in def.field_types.iter() {
        // get rid of parsing please
        if let Ok(type_) = map_string_to_cql_type(field_type) {
            do_with_referenced_udts(&mut decrement_referred_udts, &type_);
        }
    }
}
```

Reply:
Are `sorted` and `scheduled` the same?

Reply:
> Are `sorted` and `scheduled` the same?

Yes, they should be the same. Updated the snippet.